minimega: fix panic

```
panic: close of nil channel

goroutine 35577 [running]:
panic(0xac0460, 0xc82a9be9c0)
        /opt/go/src/runtime/panic.go:464 +0x3e6
ipmac.(*IPMacLearner).DelMac(0xc82ab5c700, 0xc82010d860, 0x11)
        /minimega/src/ipmac/ipmac.go:71 +0xbf
main.(*ContainerVM).launch.func1(0xc822a9c9a0, 0xc822849180, 0xc823ce56c0, 0x17)
        /minimega/src/minimega/container.go:1009 +0xb20
created by main.(*ContainerVM).launch
        /minimega/src/minimega/container.go:1028 +0x2cf2
```

--- a/src/bridge/bridges.go
+++ b/src/bridge/bridges.go
@@ -83,21 +83,27 @@ func (b Bridges) newBridge(name string) error {
 
 	br.preExist = !created
 
-	// Bring the interface up
-	if err := upInterface(br.Name, false); err != nil {
+	// Bring the interface up, start MAC <-> IP learner
+	err = upInterface(br.Name, false)
+	if err == nil {
+		err = br.startIML()
+	}
+
+	// No errors... bridge ready for use
+	if err == nil {
+		b.bridges[br.Name] = br
+		return nil
+	}
+
+	// Try to delete the bridge, if we created it
+	if created {
 		if err := ovsDelBridge(br.Name); err != nil {
 			// Welp, we're boned
 			log.Error("zombie bridge -- %v %v", br.Name, err)
 		}
-
-		return err
 	}
 
-	br.startIML()
-
-	b.bridges[br.Name] = br
-
-	return nil
+	return err
 }
 
 // Names returns a list of all the managed bridge names.

--- a/src/bridge/container.go
+++ b/src/bridge/container.go
@@ -30,7 +30,7 @@ func (b *Bridge) CreateContainerTap(t string, ns, mac string, vlan, index int) (
 	// Clean up the tap we just created, if it didn't already exist.
 	defer func() {
 		if err != nil {
-			if err := destroyVeth(tap); err != nil {
+			if err := destroyTap(tap); err != nil {
 				// Welp, we're boned
 				log.Error("zombie tap -- %v %v", tap, err)
 			}

--- a/src/bridge/ip.go
+++ b/src/bridge/ip.go
@@ -119,37 +119,9 @@ func DestroyTap(name string) error {
 func destroyTap(name string) error {
 	log.Info("destroying tuntap: %v", name)
 
-	out, err := processWrapper("ip", "tuntap", "del", "mode", "tap", name)
+	out, err := processWrapper("ip", "link", "del", name)
 	if err != nil {
 		return fmt.Errorf("destroy tap failed: %v: %v", err, out)
-	}
-
-	return nil
-}
-
-// destroyVeth destroys a veth device.
-func destroyVeth(name string) error {
-	// TODO: Needed?
-	if err := downInterface(name); err != nil {
-		return err
-	}
-
-	log.Info("destroying veth: %v", name)
-
-	args := []string{
-		"ip",
-		"link",
-		"del",
-		name,
-		"type",
-		"veth",
-		"peer",
-		"eth0",
-	}
-
-	out, err := processWrapper(args...)
-	if err != nil {
-		return fmt.Errorf("destroy veth failed: %v: %v", err, out)
 	}
 
 	return nil

--- a/src/bridge/tap.go
+++ b/src/bridge/tap.go
@@ -152,7 +152,7 @@ func (b *Bridge) RemoveTap(tap string) error {
 }
 
 // startIML starts the MAC listener for this bridge.
-func (b *Bridge) startIML() {
+func (b *Bridge) startIML() error {
 	// use openflow to redirect arp and icmp6 traffic to the local tap
 	filters := []string{
 		"dl_type=0x0806,actions=local,normal",
@@ -161,18 +161,17 @@ func (b *Bridge) startIML() {
 
 	for _, filter := range filters {
 		if err := b.addOpenflow(filter); err != nil {
-			log.Error("cannot start ip learner on bridge: %v", err)
-			return
+			return fmt.Errorf("start ip learner failed: %v", err)
 		}
 	}
 
 	iml, err := ipmac.NewLearner(b.Name)
 	if err != nil {
-		log.Error("cannot start ip learner on bridge: %v", err)
-		return
+		return fmt.Errorf("start ip learner failed: %v", err)
 	}
 
 	b.IPMacLearner = iml
+	return nil
 }
 
 // addOpenflow adds an openflow rule to the bridge using `ovs-ofctl`.

--- a/src/bridge/tap.go
+++ b/src/bridge/tap.go
@@ -128,10 +128,6 @@ func (b *Bridge) destroyTap(t string) error {
 
 	tap.Defunct = true
 
-	if tap.Container {
-		return destroyVeth(tap.Name)
-	}
-
 	return destroyTap(tap.Name)
 }
 

--- a/src/minimega/container.go
+++ b/src/minimega/container.go
@@ -857,39 +857,8 @@ func (vm *ContainerVM) launch() error {
 	// side of the network namespace. That means that unlike kvm vms, we MUST
 	// create/destroy taps on launch/kill boundaries (kvm destroys taps on
 	// flush).
-
-	// create and add taps if we are associated with any networks
-	// expose the network namespace to iptool
-	if err = vm.symlinkNetns(); err != nil {
-		log.Error("symlinkNetns: %v", err)
-	}
-
-	if err == nil {
-		for i := range vm.Networks {
-			net := &vm.Networks[i]
-
-			br, err := bridges.Get(net.Bridge)
-			if err != nil {
-				log.Error("get bridge: %v", err)
-				break
-			}
-
-			net.Tap, err = br.CreateContainerTap(net.Tap, vm.netns, net.MAC, net.VLAN, i)
-			if err != nil {
-				break
-			}
-
-			updates := make(chan ipmac.IP)
-			go vm.macSnooper(net, updates)
-
-			br.AddMac(net.MAC, updates)
-		}
-	}
-
-	if err == nil && len(vm.Networks) > 0 {
-		if err = vm.writeTaps(); err != nil {
-			log.Errorln(err)
-		}
+	if err = vm.launchNetwork(); err != nil {
+		log.Errorln(err)
 	}
 
 	childSync1.Close()
@@ -1002,7 +971,7 @@ func (vm *ContainerVM) launch() error {
 		vm.unlinkNetns()
 
 		for _, net := range vm.Networks {
-			br, err := bridges.Get(net.Bridge)
+			br, err := getBridge(net.Bridge)
 			if err != nil {
 				log.Error("get bridge: %v", err)
 			} else {
@@ -1026,6 +995,41 @@ func (vm *ContainerVM) launch() error {
 			killAck <- vm.ID
 		}
 	}()
+
+	return nil
+}
+
+func (vm *ContainerVM) launchNetwork() error {
+	// create and add taps if we are associated with any networks
+	// expose the network namespace to iptool
+	if err := vm.symlinkNetns(); err != nil {
+		return fmt.Errorf("symlinkNetns: %v", err)
+	}
+
+	for i := range vm.Networks {
+		nic := &vm.Networks[i]
+
+		br, err := getBridge(nic.Bridge)
+		if err != nil {
+			return fmt.Errorf("get bridge: %v", err)
+		}
+
+		nic.Tap, err = br.CreateContainerTap(nic.Tap, vm.netns, nic.MAC, nic.VLAN, i)
+		if err != nil {
+			return fmt.Errorf("create tap: %v", err)
+		}
+
+		updates := make(chan ipmac.IP)
+		go vm.macSnooper(nic, updates)
+
+		br.AddMac(nic.MAC, updates)
+	}
+
+	if len(vm.Networks) > 0 {
+		if err := vm.writeTaps(); err != nil {
+			return fmt.Errorf("write taps: %v", err)
+		}
+	}
 
 	return nil
 }

